### PR TITLE
Change graphic order with shorcut

### LINKED ISSUE #682: Shortcut for changing graphic order
I hope binding the same key with google slides (ctrl+page up or down I guess?)

_Originally posted by @namse in https://github.com/NamseEnt/namseent/issues/681#issuecomment-1721048558_

### PR DESCRIPTION
**Pull request by bigfoodK**
Closes #682

### CODE REVIEW

File: luda-editor/client/src/pages/sequence_edit_page/components/wysiwyg_editor/mod.rs [hunk 1]
```diff
@@ -114,51 +191,8 @@ impl Component for WysiwygEditor<'_> {
...
-                                        set_new_dragging();
-                                        mutate_dragging();
-                                        event.stop_propagation();
-                                    }
-                                    _ => {}
-                                }
-                            }
-                        }
+                        let _ = handle_graphic_order_change_shortcut(&event)
+                            || handle_graphic_move_shortcut(&event);
```
Comment (namse) [inline luda-editor/client/src/pages/sequence_edit_page/components/wysiwyg_editor/mod.rs:195]: What the hell..... seriously?

**Review by namse — CHANGES_REQUESTED** [review]

File: luda-editor/client/src/pages/sequence_edit_page/components/wysiwyg_editor/mod.rs [hunk 2]
```diff
@@ -82,6 +84,81 @@ impl Component for WysiwygEditor<'_> {
...
+                (true, false, Code::ArrowDown) => {
+                    send_backward(screen_graphics, cut_id, graphic_index)
+                }
+                (true, true, Code::ArrowDown) => {
+                    send_to_back(screen_graphics, cut_id, graphic_index)
+                }
+                _ => {
+                    return false;
+                }
+            }
```
Comment (namse) [inline luda-editor/client/src/pages/sequence_edit_page/components/wysiwyg_editor/mod.rs:111]: Why need shift?

File: luda-editor/client/src/pages/sequence_edit_page/components/wysiwyg_editor/mod.rs (hunk 1 above)
Comment (namse) [reply luda-editor/client/src/pages/sequence_edit_page/components/wysiwyg_editor/mod.rs:195]: Just create 'on_key_down' closure and make a match statement there. don't make this way which is difficult to understand your intent.

File: luda-editor/client/src/pages/sequence_edit_page/components/wysiwyg_editor/mod.rs (hunk 2 above)
Comment (namse) [reply luda-editor/client/src/pages/sequence_edit_page/components/wysiwyg_editor/mod.rs:111]: ah shift for most front or most back. ok.

**Comment by namse** [comment]
Your code could be like this,

```rust

        let handle_graphic_order_change_shortcut = |event: &KeyboardEvent| {
            let Some(graphic_index) = *editing_graphic_index else {
                return false;
            };
            if !namui::keyboard::ctrl_press() {
                return false;
            }

            match (namui::keyboard::shift_press(), event.code) {
                (true, Code::ArrowUp) => {
                    bring_to_front(screen_graphics, cut_id, graphic_index);
                    true
                }
                (false, Code::ArrowUp) => {
                    bring_forward(screen_graphics, graphic_index, cut_id);
                    true
                }
                (false, Code::ArrowDown) => {
                    send_backward(screen_graphics, cut_id, graphic_index);
                    true
                }
                (true, Code::ArrowDown) => {
                    send_to_back(screen_graphics, cut_id, graphic_index);
                    true
                }
                _ => false,
            }
        };

        let handle_graphic_move_shortcut = |event: &KeyboardEvent| {
            if !cut_editor_focused {
                return false;
            }

            let Some(graphic_index) = *editing_graphic_index else {
                return false;
            };

            let Ok(moving_with) = MovingWith::try_from(event.code) else {
                return false;
            };

            let set_new_dragging = || {
                set_dragging.set(Some(Dragging::Mover {
                    context: MoverDraggingContext {
                        start_global_xy: Xy::zero(),
                        end_global_xy: Xy::zero(),
                        container_wh: wh,
                        moving_with,
                    },
                }));
            };
            let mutate_dragging = || {
                let mut delta_xy = moving_with.delta_xy();
                if !namui::keyboard::shift_press() {
                    delta_xy *= 10.0;
                }
                set_dragging.mutate(move |dragging| {
                    if let Some(Dragging::Mover { context }) = dragging {
                        context.end_global_xy = context.end_global_xy + delta_xy;
                    }
                });
            };

            match *dragging {
                Some(Dragging::Mover { context }) => {
                    if context.moving_with.key_changed(event.code) {
                        update_sequence_with_mover(context, graphic_index);
                        set_new_dragging();
                    }
                    mutate_dragging();
                    event.stop_propagation();
                    true
                }
                None => {
                    set_new_dragging();
                    mutate_dragging();
                    event.stop_propagation();
                    true
                }
                _ => false,
            }
        };
```

File: luda-editor/client/src/pages/sequence_edit_page/components/wysiwyg_editor/change_graphic_order.rs [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+            cut_id,
+            CutUpdateAction::ChangeGraphicOrder {
+                graphic_index,
+                after_graphic_index: Some(next_or_last_graphic_index),
+            },
+        )
+    });
+}
+
+pub fn bring_forward(screen_graphics: &[(Uuid, ScreenGraphic)], graphic_index: Uuid, cut_id: Uuid) {
```
Comment (namse) [inline luda-editor/client/src/pages/sequence_edit_page/components/wysiwyg_editor/change_graphic_order.rs:49]: This function have only different parameter order unlike other functions. graphic_index and cut_id

File: luda-editor/client/src/pages/sequence_edit_page/components/wysiwyg_editor/mod.rs (hunk 1 above)
Comment (namse) [reply luda-editor/client/src/pages/sequence_edit_page/components/wysiwyg_editor/mod.rs:195]: Nah, I have no better idea for this.

Comment (bigfoodK) [reply luda-editor/client/src/pages/sequence_edit_page/components/wysiwyg_editor/mod.rs:195]: I have better options.
I think option 2 looks fine.

## option 1. more literal than original
```rust
if handle_graphic_order_change_shortcut(&event) {
} else if handle_graphic_move_shortcut(&event) {
}
```

## option 2. more kind option than option 1
```rust
Event::KeyDown { event } => 'handle_key_down: {
    let handle_graphic_order_change_shortcut_executed =
        handle_graphic_order_change_shortcut(&event);
    if handle_graphic_order_change_shortcut_executed {
        break 'handle_key_down;
    }

    let handle_graphic_move_shortcut_executed =
        handle_graphic_move_shortcut(&event);
    if handle_graphic_move_shortcut_executed {
        break 'handle_key_down;
    }
}
```

File: luda-editor/client/src/pages/sequence_edit_page/components/wysiwyg_editor/change_graphic_order.rs (hunk 3 above)
Comment (bigfoodK) [reply luda-editor/client/src/pages/sequence_edit_page/components/wysiwyg_editor/change_graphic_order.rs]: What? How could that happen?

Comment (bigfoodK) [reply luda-editor/client/src/pages/sequence_edit_page/components/wysiwyg_editor/change_graphic_order.rs]: Fixed with 92c707f

**Review by namse — APPROVED** [review]
I hope you write better quality code, or bad code will come to you when you are really busy and choke you.

**Pull request merged**